filebeat/scripts/generator/fields/main.go

```diff
@@ -0,0 +1,340 @@
+			pp, err := accumulatePatterns(ee)
+			if err != nil {
+				return nil, err
+			}
+			patterns = append(patterns, pp...)
+		}
+		if rm, ok := e["remove"]; ok {
+			rmFields = accumulateRemoveFields(rm, rmFields)
+		}
+		if mv, ok := e["rename"]; ok {
+			mvFields = accumulateRenameFields(mv, mvFields)
+		}
+	}
+
+	if patterns == nil {
+		return nil, fmt.Errorf("No patterns in pipeline")
+	}
+
+	return &processors{
+		patterns: patterns,
+		remove:   rmFields,
+		rename:   mvFields,
+	}, nil
+}
+
+func getFieldByName(f []*fieldYml, name string) *fieldYml {
+	for _, ff := range f {
+		if ff.Name == name {
+			return ff
+		}
+	}
+	return nil
+}
+
+func insertLastField(f []*fieldYml, name, typeName string, noDoc bool) []*fieldYml {
+	ff := getFieldByName(f, name)
+	if ff != nil {
+		return f
+	}
+
+	nf := newFieldYml(name, types[typeName], noDoc)
+	return append(f, nf)
+}
+
+func insertGroup(out []*fieldYml, field field, index, count int, noDoc bool) []*fieldYml {
+	g := getFieldByName(out, field.Elements[index])
+	if g != nil {
+		g.Fields = generateField(g.Fields, field, index+1, count, noDoc)
+		return out
+	}
+
+	var groupFields []*fieldYml
+	groupFields = generateField(groupFields, field, index+1, count, noDoc)
+	group := newFieldYml(field.Elements[index], "group", noDoc)
+	group.Fields = groupFields
+	return append(out, group)
+}
+
+func generateField(out []*fieldYml, field field, index, count int, noDoc bool) []*fieldYml {
+	if index+1 == count {
+		return insertLastField(out, field.Elements[index], field.Type, noDoc)
+	}
+	return insertGroup(out, field, index, count, noDoc)
+}
+
+func generateFields(f []field, noDoc bool) []*fieldYml {
+	var out []*fieldYml
+	for _, ff := range f {
+		out = generateField(out, ff, 1, len(ff.Elements), noDoc)
+	}
+	return out
+}
+
+func (p *pipeline) toFieldsYml(noDoc bool) ([]byte, error) {
+	pr, err := getProcessors(p.Processors)
+	if err != nil {
+		return nil, err
+	}
+
+	var fs []field
+	fs, err = pr.processFields()
+	if err != nil {
+		return nil, err
+	}
+
+	f := generateFields(fs, noDoc)
+	var d []byte
+	d, err = yaml.Marshal(&f)
+
+	return d, nil
+}
+
+func writeFieldsYml(beatsPath, module, fileset string, f []byte) error {
+	p := fmt.Sprintf(fieldsYmlPath, beatsPath, module, fileset)
+	err := ioutil.WriteFile(p, f, 0664)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func main() {
+	module := flag.String("module", "", "Name of the module")
+	fileset := flag.String("fileset", "", "Name of the fileset")
+	beatsPath := flag.String("beats_path", ".", "Path to elastic/beats")
+	noDoc := flag.Bool("nodoc", false, "Generate documentation for fields")
+	flag.Parse()
+
+	if *module == "" {
+		fmt.Println("Missing parameter: module")
+		os.Exit(1)
+	}
+
+	if *fileset == "" {
+		fmt.Println("Missing parameter: fileset")
+		os.Exit(1)
+	}
+
+	p, err := readPipeline(*beatsPath, *module, *fileset)
+	if err != nil {
+		fmt.Printf("Cannot read pipeline.yml of fileset: %v\n", err)
+		os.Exit(2)
```

Comment:
I stumbled a bit over the part that you keep increasing the Exit codes, but as long as it's not 0, should work.